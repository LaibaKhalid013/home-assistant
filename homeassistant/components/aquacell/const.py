"""Constants for the Aquacell integration."""

DOMAIN = "aquacell"
DATA_AQUACELL = "DATA_AQUACELL"
