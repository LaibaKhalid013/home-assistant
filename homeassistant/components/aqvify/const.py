"""Constants for the Aqvify integration."""

import logging

DOMAIN = "aqvify"

CONF_DEVICE_KEY = "device_key"

LOGGER = logging.getLogger(__package__)
