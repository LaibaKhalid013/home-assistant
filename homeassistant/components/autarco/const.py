"""Constants for the Autarco integration."""

from __future__ import annotations

from datetime import timedelta
import logging
from typing import Final

DOMAIN: Final = "autarco"
LOGGER = logging.getLogger(__package__)
SCAN_INTERVAL = timedelta(minutes=1)

CONF_PUBLIC_KEY: Final = "public_key"

SENSORS_SOLAR: Final = "solar"
