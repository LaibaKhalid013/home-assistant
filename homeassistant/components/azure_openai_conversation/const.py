"""Constants for the OpenAI Conversation integration."""

import logging

DOMAIN = "azure_openai_conversation"
LOGGER = logging.getLogger(__package__)

CONF_AZURE_OPENAI_RESOURCE = "azure_openai_resource"
CONF_RECOMMENDED = "recommended"
CONF_PROMPT = "prompt"
CONF_CHAT_MODEL = "chat_model"
RECOMMENDED_CHAT_MODEL = "gpt-4"
CONF_MAX_TOKENS = "max_tokens"
RECOMMENDED_MAX_TOKENS = 150
CONF_TOP_P = "top_p"
RECOMMENDED_TOP_P = 1.0
CONF_TEMPERATURE = "temperature"
RECOMMENDED_TEMPERATURE = 1.0
AZURE_OPEN_API_VERSION = "2024-02-01"
