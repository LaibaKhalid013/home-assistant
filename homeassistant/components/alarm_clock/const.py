"""Constants for the Alarm Clock integration."""

import voluptuous as vol

from homeassistant.const import CONF_ICON, CONF_NAME
import homeassistant.helpers.config_validation as cv

DOMAIN = "alarm_clock"
ENTITY_ID_FORMAT = DOMAIN + ".{}"

ATTR_ALARM_TIME = "alarm_time"
ATTR_REPEAT_DAYS = "repeat_days"
ATTR_NEXT_ALARM = "next_alarm"

CONF_ALARM_TIME = "alarm_time"
CONF_REPEAT_DAYS = "repeat_days"

EVENT_ALARM_CLOCK_STARTED = "alarm_clock.started"
EVENT_ALARM_CLOCK_FINISHED = "alarm_clock.finished"
EVENT_ALARM_CLOCK_CANCELLED = "alarm_clock.cancelled"
EVENT_ALARM_CLOCK_CHANGED = "alarm_clock.changed"

STORAGE_KEY = DOMAIN
STORAGE_VERSION = 1

STORAGE_FIELDS = {
    vol.Required(CONF_NAME): cv.string,
    vol.Optional(CONF_ICON): cv.icon,
    vol.Required(CONF_ALARM_TIME): cv.time,
    vol.Required(CONF_REPEAT_DAYS): cv.weekdays,
}
