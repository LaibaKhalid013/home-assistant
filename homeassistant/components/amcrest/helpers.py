"""Helpers for amcrest component."""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from .const import DOMAIN

if TYPE_CHECKING:
    from homeassistant.helpers.typing import UndefinedType


def service_signal(service: str, *args: str) -> str:
    """Encode signal."""
    return "_".join([DOMAIN, service, *args])


def log_update_error(
    logger: logging.Logger,
    action: str,
    name: str | UndefinedType | None,
    entity_type: str,
    error: Exception,
    level: int = logging.ERROR,
) -> None:
    """Log an update error."""
    logger.log(
        level,
        "Could not %s %s %s due to error: %s",
        action,
        name,
        entity_type,
        error.__class__.__name__,
    )
