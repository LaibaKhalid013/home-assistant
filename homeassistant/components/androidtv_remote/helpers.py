"""Helper functions for Android TV Remote integration."""

from __future__ import annotations

from typing import TYPE_CHECKING

from androidtvremote2 import AndroidTVRemote

from homeassistant.helpers.storage import STORAGE_DIR

from .const import CONF_ENABLE_IME, CONF_ENABLE_IME_DEFAULT_VALUE

if TYPE_CHECKING:
    from homeassistant.config_entries import ConfigEntry
    from homeassistant.core import HomeAssistant


def create_api(hass: HomeAssistant, host: str, enable_ime: bool) -> AndroidTVRemote:
    """Create an AndroidTVRemote instance."""
    return AndroidTVRemote(
        client_name="Home Assistant",
        certfile=hass.config.path(STORAGE_DIR, "androidtv_remote_cert.pem"),
        keyfile=hass.config.path(STORAGE_DIR, "androidtv_remote_key.pem"),
        host=host,
        loop=hass.loop,
        enable_ime=enable_ime,
    )


def get_enable_ime(entry: ConfigEntry) -> bool:
    """Get value of enable_ime option or its default value."""
    return entry.options.get(CONF_ENABLE_IME, CONF_ENABLE_IME_DEFAULT_VALUE)
