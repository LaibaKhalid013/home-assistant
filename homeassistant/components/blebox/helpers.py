"""Blebox helpers."""

from __future__ import annotations

from typing import TYPE_CHECKING

import aiohttp

from homeassistant.helpers.aiohttp_client import (
    async_create_clientsession,
    async_get_clientsession,
)

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant


def get_maybe_authenticated_session(
    hass: HomeAssistant, password: str | None, username: str | None
) -> aiohttp.ClientSession:
    """Return proper session object."""
    if username and password:
        auth = aiohttp.BasicAuth(login=username, password=password)
        return async_create_clientsession(hass, auth=auth)

    return async_get_clientsession(hass)
