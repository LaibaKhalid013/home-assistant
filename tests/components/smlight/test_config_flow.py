"""Test the SMLIGHT SLZB config flow."""

from ipaddress import ip_address
from json import loads
from unittest.mock import AsyncMock, patch

from pysmlight.exceptions import SmlightAuthError, SmlightConnectionError, SmlightError
from pysmlight.web import Info
from syrupy.assertion import SnapshotAssertion

from homeassistant import config_entries
from homeassistant.components import zeroconf
from homeassistant.components.smlight.const import DOMAIN
from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME
from homeassistant.core import HomeAssistant
from homeassistant.data_entry_flow import FlowResultType
from homeassistant.helpers.device_registry import format_mac

from tests.common import MockConfigEntry, load_fixture

DISCOVERY_INFO = zeroconf.ZeroconfServiceInfo(
    ip_address=ip_address("127.0.0.1"),
    ip_addresses=[ip_address("127.0.0.1")],
    hostname="SLZB-06.local.",
    name="mock_name",
    port=6638,
    properties={"mac": "AA:BB:CC:DD:EE:FF"},
    type="mock_type",
)

DISCOVERY_INFO_LEGACY = zeroconf.ZeroconfServiceInfo(
    ip_address=ip_address("127.0.0.1"),
    ip_addresses=[ip_address("127.0.0.1")],
    hostname="SLZB-06.local.",
    name="mock_name",
    port=6638,
    properties={},
    type="mock_type",
)


async def test_user_flow(hass: HomeAssistant, mock_setup_entry: AsyncMock) -> None:
    """Test the full manual user flow."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )
    assert result.get("type") == FlowResultType.FORM
    assert result.get("step_id") == "user"
    assert result.get("errors") == {}

    mock_info = Info.from_dict(loads(load_fixture("smlight/info.json")))
    with (
        patch(
            "pysmlight.web.Api2.get_info",
            return_value=mock_info,
        ),
        patch(
            "pysmlight.web.Api2.check_auth_needed",
            return_value=False,
        ),
    ):
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                CONF_HOST: "slzb-06.local",
            },
        )
        await hass.async_block_till_done()

    assert result["type"] == FlowResultType.CREATE_ENTRY
    assert result["title"] == "SLZB-06p7"
    assert result["data"] == {
        CONF_HOST: "slzb-06.local",
    }
    assert len(mock_setup_entry.mock_calls) == 1


async def test_zeroconf_flow(
    hass: HomeAssistant,
    mock_setup_entry: AsyncMock,
    snapshot: SnapshotAssertion,
) -> None:
    """Test the full zeroconf flow including authentication."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_ZEROCONF}, data=DISCOVERY_INFO
    )

    assert result.get("description_placeholders") == {"host": "SLZB-06.local"}
    assert result.get("type") is FlowResultType.FORM
    assert result.get("step_id") == "confirm_discovery"

    mock_info = Info.from_dict(loads(load_fixture("smlight/info.json")))
    with (
        patch(
            "pysmlight.web.Api2.get_info",
            return_value=mock_info,
        ),
        patch(
            "pysmlight.web.Api2.check_auth_needed",
            return_value=True,
        ),
        patch(
            "pysmlight.web.Api2.authenticate",
        ),
    ):
        progress = hass.config_entries.flow.async_progress()
        assert len(progress) == 1
        assert progress[0].get("flow_id") == result["flow_id"]
        assert progress[0]["context"].get("confirm_only") is True

        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"], user_input={}
        )

        assert result2.get("type") is FlowResultType.FORM
        assert result2.get("step_id") == "auth"

        progress2 = hass.config_entries.flow.async_progress()
        assert len(progress2) == 1
        assert progress2[0].get("flow_id") == result["flow_id"]

        result3 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            user_input={
                CONF_USERNAME: "test-user",
                CONF_PASSWORD: "test-pass",
            },
        )

        assert result3.get("type") is FlowResultType.CREATE_ENTRY
        assert result3 == snapshot

        assert len(mock_setup_entry.mock_calls) == 1


async def test_user_form(hass: HomeAssistant, mock_setup_entry: AsyncMock) -> None:
    """Test we get the user form."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )

    assert result.get("type") == FlowResultType.FORM
    assert result.get("step_id") == "user"


async def test_user_device_exists_abort(
    hass: HomeAssistant, mock_config_entry: MockConfigEntry
) -> None:
    """Test we abort user flow if device already configured."""
    mock_info = Info.from_dict(loads(load_fixture("smlight/info.json")))
    with (
        patch(
            "pysmlight.web.Api2.get_info",
            return_value=mock_info,
        ),
        patch(
            "pysmlight.web.Api2.check_auth_needed",
            return_value=False,
        ),
    ):
        mock_config_entry.add_to_hass(hass)
        result = await hass.config_entries.flow.async_init(
            DOMAIN,
            context={"source": config_entries.SOURCE_USER},
            data={
                CONF_HOST: "slzb-06.lan",
            },
        )

        assert result.get("type") is FlowResultType.ABORT
        assert result.get("reason") == "already_configured"


async def test_zeroconf_device_exists_abort(
    hass: HomeAssistant, mock_config_entry: MockConfigEntry
) -> None:
    """Test we abort zeroconf flow if device already configured."""
    mock_info = Info.from_dict(loads(load_fixture("smlight/info.json")))
    with (
        patch(
            "pysmlight.web.Api2.get_info",
            return_value=mock_info,
        ),
        patch(
            "pysmlight.web.Api2.check_auth_needed",
            return_value=False,
        ),
    ):
        mock_config_entry.add_to_hass(hass)
        result = await hass.config_entries.flow.async_init(
            DOMAIN,
            context={"source": config_entries.SOURCE_ZEROCONF},
            data=DISCOVERY_INFO,
        )

        assert result.get("type") is FlowResultType.ABORT
        assert result.get("reason") == "already_configured"


async def test_user_invalid_auth(
    hass: HomeAssistant, mock_setup_entry: AsyncMock
) -> None:
    """Test we handle invalid auth."""
    with (
        patch(
            "pysmlight.web.Api2.check_auth_needed",
            return_value=True,
        ),
        patch(
            "pysmlight.web.Api2.authenticate",
            side_effect=SmlightAuthError,
        ),
    ):
        result = await hass.config_entries.flow.async_init(
            DOMAIN,
            context={"source": config_entries.SOURCE_USER},
            data={
                CONF_HOST: "slzb-06.local",
            },
        )

        assert result["type"] == FlowResultType.FORM
        assert result.get("step_id") == "auth"

        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                CONF_USERNAME: "test",
                CONF_PASSWORD: "bad",
            },
        )

        assert result2["type"] == FlowResultType.FORM
        assert result2["errors"] == {"base": "invalid_auth"}
        assert result2.get("step_id") == "auth"


async def test_user_api_exception(
    hass: HomeAssistant, mock_setup_entry: AsyncMock
) -> None:
    """Test we handle unknown exceptions in pysmlight api."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )

    with patch(
        "pysmlight.web.Api2.check_auth_needed",
        side_effect=SmlightError,
    ):
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                CONF_HOST: "slzb-06.local",
            },
        )

        assert result["type"] == FlowResultType.FORM
        assert result["errors"] == {"base": "unknown"}
        assert result.get("step_id") == "user"


async def test_user_cannot_connect(
    hass: HomeAssistant, mock_setup_entry: AsyncMock
) -> None:
    """Test we handle user cannot connect error."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )

    with patch(
        "pysmlight.web.Api2.check_auth_needed",
        side_effect=SmlightConnectionError,
    ):
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                CONF_HOST: "slzb-06.local",
            },
        )

        assert result["type"] == FlowResultType.FORM
        assert result["errors"] == {"base": "cannot_connect"}
        assert result.get("step_id") == "user"


async def test_zeroconf_cannot_connect(
    hass: HomeAssistant, mock_setup_entry: AsyncMock
) -> None:
    """Test we handle zeroconf cannot connect error."""
    with patch(
        "pysmlight.web.Api2.check_auth_needed",
        side_effect=SmlightConnectionError,
    ):
        result = await hass.config_entries.flow.async_init(
            DOMAIN,
            context={"source": config_entries.SOURCE_ZEROCONF},
            data=DISCOVERY_INFO,
        )
        assert result["type"] == FlowResultType.FORM
        assert result.get("step_id") == "confirm_discovery"

        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {},
        )

        assert result2["type"] == FlowResultType.FORM
        assert result2["errors"] == {"base": "cannot_connect"}
        assert result2.get("step_id") == "confirm_discovery"


async def test_zeroconf_legacy_mac(
    hass: HomeAssistant, mock_setup_entry: AsyncMock
) -> None:
    """Test we can get unique id MAC address for older firmwares."""
    mock_info = Info.from_dict(loads(load_fixture("smlight/info.json")))
    with (
        patch(
            "pysmlight.web.Api2.get_info",
            return_value=mock_info,
        ),
        patch(
            "pysmlight.web.Api2.check_auth_needed",
            return_value=False,
        ),
    ):
        result = await hass.config_entries.flow.async_init(
            DOMAIN,
            context={"source": config_entries.SOURCE_ZEROCONF},
            data=DISCOVERY_INFO_LEGACY,
        )

        progress = hass.config_entries.flow.async_progress()
        assert len(progress) == 1
        assert "context" in progress[0]
        assert progress[0]["context"].get("unique_id") == format_mac(mock_info.MAC)
        assert progress[0].get("flow_id") == result["flow_id"]
        assert result.get("description_placeholders") == {"host": "SLZB-06.local"}


async def test_reauth_flow(
    hass: HomeAssistant, mock_config_entry: MockConfigEntry
) -> None:
    """Test reauth flow completes successfully."""
    mock_info = Info.from_dict(loads(load_fixture("smlight/info.json")))
    with (
        patch(
            "pysmlight.web.Api2.get_info",
            return_value=mock_info,
        ),
        patch(
            "pysmlight.web.Api2.check_auth_needed",
            return_value=True,
        ),
        patch(
            "pysmlight.web.Api2.authenticate",
        ),
    ):
        mock_config_entry.add_to_hass(hass)
        result = await hass.config_entries.flow.async_init(
            DOMAIN,
            context={
                "source": config_entries.SOURCE_REAUTH,
                "entry_id": mock_config_entry.entry_id,
            },
            data=mock_config_entry.data,
        )

        assert result.get("type") == FlowResultType.FORM
        assert result.get("step_id") == "reauth_confirm"

        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                CONF_USERNAME: "test-user",
                CONF_PASSWORD: "test-pass",
            },
        )

        assert result2.get("type") == FlowResultType.ABORT
        assert result2.get("reason") == "reauth_successful"


async def test_reauth_error(
    hass: HomeAssistant, mock_config_entry: MockConfigEntry
) -> None:
    """Test reauth flow with error."""
    mock_info = Info.from_dict(loads(load_fixture("smlight/info.json")))
    with (
        patch(
            "pysmlight.web.Api2.get_info",
            return_value=mock_info,
        ),
        patch(
            "pysmlight.web.Api2.check_auth_needed",
            return_value=True,
        ),
        patch(
            "pysmlight.web.Api2.authenticate",
            side_effect=SmlightAuthError,
        ),
    ):
        mock_config_entry.add_to_hass(hass)
        result = await hass.config_entries.flow.async_init(
            DOMAIN,
            context={
                "source": config_entries.SOURCE_REAUTH,
                "entry_id": mock_config_entry.entry_id,
            },
            data=mock_config_entry.data,
        )

        assert result.get("type") == FlowResultType.FORM
        assert result.get("step_id") == "reauth_confirm"

        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                CONF_USERNAME: "test",
                CONF_PASSWORD: "bad",
            },
        )

        assert result2.get("type") == FlowResultType.ABORT
        assert result2.get("reason") == "reauth_failed"
