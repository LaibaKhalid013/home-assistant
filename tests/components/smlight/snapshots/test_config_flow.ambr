# serializer version: 1
# name: test_zeroconf_flow
  FlowResultSnapshot({
    'context': dict({
      'confirm_only': False,
      'source': 'zeroconf',
      'title_placeholders': dict({
        'name': 'SLZB-06',
      }),
      'unique_id': 'aa:bb:cc:dd:ee:ff',
    }),
    'data': dict({
      'host': 'SLZB-06.local',
      'password': 'test-pass',
      'username': 'test-user',
    }),
    'description': None,
    'description_placeholders': None,
    'flow_id': <ANY>,
    'handler': 'smlight',
    'minor_version': 1,
    'options': dict({
    }),
    'result': ConfigEntrySnapshot({
      'data': dict({
        'host': 'SLZB-06.local',
        'password': 'test-pass',
        'username': 'test-user',
      }),
      'disabled_by': None,
      'domain': 'smlight',
      'entry_id': <ANY>,
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'zeroconf',
      'title': 'SLZB-06p7',
      'unique_id': 'aa:bb:cc:dd:ee:ff',
      'version': 1,
    }),
    'title': 'SLZB-06p7',
    'type': <FlowResultType.CREATE_ENTRY: 'create_entry'>,
    'version': 1,
  })
# ---
